Sources/ArcGISToolkit/Components/FormView/FormInputs/InputFooter.swift

```diff
@@ -17,23 +17,234 @@ import SwiftUI
+                    bundle: .toolkitModule,
+                    comment: "Text indicating a field's value must be of the correct type."
+                )
+            }
+        }
+    }
+    
+    /// Text indicating a field's exact number of allowed characters.
+    /// - Note: This is intended to be used in instances where the character minimum and maximum are
+    /// identical, such as an ID field.
+    var exactLengthMessage: Text {
+        Text(
+            "Enter \(lengthRange!.lowerBound) characters",
+            bundle: .toolkitModule,
+            comment: "Text indicating the user should enter a field's exact number of required characters."
+        )
+    }
+    
+    /// Determines whether an error is showing in the footer.
+    var isShowingError: Bool {
+        element.isEditable && primaryError != nil && model.previouslyFocusedFields.contains(element)
+    }
+    
+    /// The alloable range of number of characters in the input.
+    var lengthRange: ClosedRange<Int>? {
+        if let input = element.input as? TextAreaFormInput {
+            return input.minLength...input.maxLength
+        } else if let input = element.input as? TextBoxFormInput, element.fieldType == .text {
+            return input.minLength...input.maxLength
+        } else {
+            return nil
+        }
+    }
+    
+    /// Text indicating a field's value must be within the allowed length range.
+    var lengthRangeMessage: Text {
+        Text(
+            "Enter \(lengthRange!.lowerBound) to \(lengthRange!.upperBound) characters",
+            bundle: .toolkitModule,
+            comment: """
+                         Text indicating a field's value must be within the 
+                         allowed length range. The first and second parameter
+                         hold the minimum and maximum length respectively.
+                         """
+        )
+    }
+    
+    /// Text indicating a field's maximum number of allowed characters.
+    var maximumLengthMessage: Text {
+        Text(
+            "Maximum \(lengthRange!.upperBound) characters",
+            bundle: .toolkitModule,
+            comment: "Text indicating a field's maximum number of allowed characters."
+        )
+    }
+    
+    /// Text indicating a field's maximum number of allowed characters.
+    var minimumLengthMessage: Text {
+        Text(
+            "Minimum \(lengthRange!.lowerBound) characters",
+            bundle: .toolkitModule,
+            comment: "Text indicating a field's minimum number of allowed characters."
+        )
+    }
+    
+    /// The error to display for the input. If the element is not focused and has a required error this will be
+    /// the primary error. Otherwise the primary error is the first error in the element's set of errors.
+    var primaryError: ArcGIS.FeatureFormError? {
+        let elementErrors = model.validationErrors[element.fieldName] as? [ArcGIS.FeatureFormError]
```

Comment:
Since `primaryError` is computed and not a `@State` or `@Published`, will the UI update if it changes? `element.validationErrors` is never accessed, we're just relying on `featureForm.validationErrors`, which is somewhat inefficient, as that goes through all elements.

I'm wondering if we can make the errors a `@State` property and localize them to the individual elements. Or make `primaryError` a `@State` object which contains the error to display, which is then passed into the `InputFooter`.

So after `model.evaluateExpressions()` is called in each of the input types, the `element.validationErrors` are retrieved and used to create `primaryError` (or another `@State` property).

I want to make sure that the error display is updated in response to validation errors and we're not always getting all validation errors from the feature form.

Also, I had envisioned `featureForm.evaluationErrors` to be only called in response to a `submit` action.

The one question I have is how does that work for other elements whose values are updated in response to expression evaluation, but then have validation errors? Are we showing those errors (the elements will never get focus, as they are non-editable)?

Reply:
> Since primaryError is computed and not a `@State` or `@Published`, will the UI update if it changes? 

Yes, in my testing the UI changes correctly as the validation errors change. Since `InputFooter` was redrawing whenever the `FormViewModel` published changes (which was on every character input), `featureForm.validationErrors` was also being hit and re-accessed each time.

> element.validationErrors is never accessed, we're just relying on featureForm.validationErrors, which is somewhat inefficient, as that goes through all elements. ... I want to make sure that the error display is updated in response to validation errors and we're not always getting all validation errors from the feature form. ... Also, I had envisioned featureForm.evaluationErrors to be only called in response to a submit action.

👍  I've changed this branch to access `element.validationErrors` instead.

> The one question I have is how does that work for other elements whose values are updated in response to expression evaluation, but then have validation errors? Are we showing those errors (the elements will never get focus, as they are non-editable)?

Following Kaushik's validation algorithm, non-editable fields will never show validation errors. Every other field will show validation errors either once it has or had focus or the force validation flag is set.

> I'm wondering if we can make the errors a `@State` property and localize them to the individual elements. ... So after model.evaluateExpressions() is called in each of the input types, the element.validationErrors are retrieved and used to create primaryError (or another @State property).

I've implemented this in another branch ([df/NewValidationAPI-alt1](https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/tree/df/NewValidationAPI-alt1)). The  diff between the two branches is [here](https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/compare/df/NewValidationAPI...df/NewValidationAPI-alt1). Code-wise this implementation seems more brittle and there's a lot of repetitive conditional casting to (`element.validationErrors as? [FeatureFormError] ?? []`). Efficiency-wise `element.validationErrors` is hit a much less often but I'm not sure if that outweighs the downside of having all of the repetitive code in each input. Thoughts?

Reply:
Ok, thanks for looking into that. It's kind of messy having to do the casting and the duplicate code. So keep what you have.